compiler/rustc_hir_analysis/src/astconv/mod.rs

```diff
@@ -2331,13 +2331,114 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
         self.ast_ty_to_ty_inner(ast_ty, false, true)
     }
 
+    fn check_delegation_constraints(&self, res_id: DefId, span: Span, emit: bool) -> bool {
+        #[inline]
+        fn generics_param_count(param_count: ty::GenericParamCount) -> usize {
+            param_count.lifetimes + param_count.types + param_count.consts
+        }
+
+        let mut error_occured = false;
+        let res_span = self.tcx().def_span(res_id);
+        let mut try_emit = |descr| {
+            if emit {
+                self.tcx().sess.emit_err(crate::errors::NotSupportedDelegation {
+                    span,
+                    descr,
+                    res_span,
+                });
+            }
+            error_occured = true;
+        };
+
+        if let Some(node) = self.tcx().hir().get_if_local(res_id)
+            && let Some(decl) = node.fn_decl()
+            && let hir::FnRetTy::Return(ty) = decl.output
+        {
+            match ty.kind {
+                hir::TyKind::InferDelegation(_, _) => try_emit("recursive delegation"),
+                hir::TyKind::OpaqueDef(_, _, _) => {
+                    try_emit("delegation to a function with opaque type")
+                }
+                _ => {}
+            }
+        }
+
+        let res_generics = self.tcx().generics_of(res_id);
+        let parent = self.tcx().parent(self.item_def_id());
+        let parent_generics = self.tcx().generics_of(parent);
+        let self_ty_bias = (self.tcx().def_kind(parent) == DefKind::Trait) as usize;
+        if parent_generics.count() + generics_param_count(res_generics.own_counts()) > self_ty_bias
+        {
+            try_emit("delegation with early bound generics");
+        }
+
+        if self.tcx().asyncness(res_id) == ty::Asyncness::Yes {
+            try_emit("delegation to async functions");
+        }
+
+        if self.tcx().constness(res_id) == hir::Constness::Const {
+            try_emit("delegation to const functions");
+        }
+
+        let res_sig = self.tcx().fn_sig(res_id).instantiate_identity();
+        if res_sig.c_variadic() {
+            try_emit("delegation to variadic functions");
+            return error_occured;
```

Comment:
`error_occured` is returned here, but not in other checks, not sure why.

Reply:
Comment was added.